[c10d][fr] Add logging of nccl_version into fr and its dump

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151048

Users also want to see the nccl version in the FR dump so let's add it to FR. We only add it per rank per PG nccl comm, so this is really add a couple bytes to FR memory.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[c10d][fr] Add logging of nccl_version into fr and its dump

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -24,9 +24,10 @@ namespace c10d {
 // Also update both JSON and Pickle dumps to make use of the newly defined
 // field(s).
-DEFINE_CONSTANT(version_val, "2.5")
+DEFINE_CONSTANT(version_val, "2.6")
```
Maybe use [`ncclGetVersion`](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclgetversion) rather than defining a macro/constant? (reduce maintenance load)

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -685,6 +694,7 @@ std::string FlightRecorder::dump(
   // common values
   result.insert(version_key, version_val);
```
what's the difference between this line and the `record_nccl_version` API?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2946,6 +2946,7 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
 
   FlightRecorder::get()->record_pg_ranks(
       std::make_tuple(pg_uid_, pg_desc_), groupRanks());
+  FlightRecorder::get()->record_nccl_version(std::move(getNcclVersion()));
```
nit: consider naming APIs in backend agnostic way.

fduwjj committed:
Update on "[c10d][fr] Add logging of nccl_version into fr and its dump"


Users also want to see the nccl version in the FR dump so let's add it to FR. We only add it per rank per PG nccl comm, so this is really add a couple bytes to FR memory.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -i

fduwjj commented:
@pytorchbot merge -f

fduwjj commented:
@pytorchbot merge -f "No CI failure"
